### ISSUE TITLE
Wndows: issue with escaping auto-completed filenames after :!{cmd}

### ISSUE THREAD

**Issue by lkintact** [vim/vim#5077]
1. Create a file with the name `file with spaces and a '.txt`
2. Run `gvim --clean`
3. Type `:read !type f` and press <kbd>Tab</kbd> to auto-complete the filename
4. The command line becomes `:read !type file\ with\ spaces\ and\ a\ \'.txt
`
5. Press <kbd>Enter</kbd>. "shell returned 1" message is displayed and an error emitted by the "type" command is put into the buffer.

Apparently this happens because "type" expects `"file with spaces and a '.txt"`, not `file\ with\ spaces\ and\ a\ \'.txt` as the filename parameter (as perhaps any Windows command-line tool would). I tried searching through the docs for a solution, but didn't find anything, so maybe this should be considered a bug?

Vim: https://github.com/vim/vim-win32-installer/releases/download/v8.1.2168/gvim_8.1.2168_x86_signed.zip,
Windows 10 Home 1803.

**Comment by chrisbra** [comment]
I don't think this is a bug. Most vim commands expect an escaped filename, so that is what `<Tab>` does complete. Trying to make it complete depending on how the command would expect it is probably not worth the effort (it could still be wrong, because the user edits the command in front of it later)

**Comment by lkintact** [comment]
@chrisbra  Unfortunately with the current situation using `<Tab>` completion after `:!{cmd}` (at least on Windows, not sure about other OSes) is of very little use in case the filename has lots of spaces and other special chars, because for the command to work properly the user is required to manually remove all the backslashes and doublequote the filename.

Maybe what `<Tab>` completes should depend on whether `:!{cmd}` precedes the place where <kbd>Tab</kbd> was pressed and what OS Vim runs under?

**Comment by chrisbra** [comment]
If the filename has spaces, it is correct to escape it. Think of using `:e` and use `tab` to find the correct file. And changing `tab` to behave differently would be a backwards incompatible change. 
What you can do instead is, use the command line window (e.g. press `<c-f>` in command line mode, see `:h c_CTRL-F`, edit the command there and use `Ctrl-X Ctrl-F` to complete the filepath there (which does not escape the path) and press enter to execute the command). 

Perhaps it would be a good idea to make the file path completion of `Ctrl-X Ctrl-F` available in command line mode. Not sure what would be a good key for that, we are running out of ctrl keys for that. Looks like we still have `c_CTRL-O`, `c_CTRL-S` (that one is troublesome in terminals), `c_CTRL-X` and `c_CTRL-Z` (which is also troublesome). Perhaps using `c_CTRL-X` and then use whatever completion you would use in insert mode (e.g. `CTRL-X CTRL-F` would expand file paths literally)?

**Comment by Houl** [comment]
So you cannot make `<Tab>` work correctly after `!`?

**Comment by chrisbra** [comment]
> So you cannot make <Tab> work correctly after !?

It depends on what you define as `correct`. It is highly dependent on whatever external command you run and how it expects its paths. It could expect spaces to be escaped or not, I don't think there is a general solution to this problem.

**Comment by lkintact** [comment]
> Perhaps using `c_CTRL-X` and then use whatever completion you would use in insert mode (e.g. `CTRL-X CTRL-F` would expand file paths literally)?

Sounds great to me, but of course I don't have a say in this.
At any rate, thank you very much for the command line window solution.
